src/function/mtk.jl

```diff
@@ -0,0 +1,60 @@
+struct AutoModelingToolkit <: AbstractADType end
+
+function instantiate_function(f, x, ::AutoModelingToolkit, p, num_cons=0)
+
+    sys = ModelingToolkit.modelingtoolkitize(OptimizationProblem(f, x, p))
+    println(sys)
+    if f.grad === nothing
+        grad_oop, grad_iip = ModelingToolkit.generate_gradient(sys, expression=Val{false})
+        grad(J, u) = (grad_iip(J, u, p); J)
+    else
+        grad = f.grad
+    end
+
+    if f.hess === nothing
+        hess_oop, hess_iip = ModelingToolkit.generate_hessian(sys, expression=Val{false})
+        hess(J, u) = (hess_iip(J, u, p); J)
+    else
+        hess = f.hess
+    end
+
+    if f.hv === nothing
+        hv = function (H, θ, v, args...)
+            res = ArrayInterface.zeromatrix(θ)
+            hess(res, θ, args...)
+            H .= res * v
+        end
+    else
+        hv = f.hv
+    end
+
+    if f.cons === nothing
+        cons = nothing
+    else
+        cons = (θ) -> f.cons(θ, p)
+        cons_sys = ModelingToolkit.modelingtoolkitize(OptimizationProblem(f.cons, x, p); checks=false)
```

Comment:
Does this work? The constraints are supposed to be a vector function. The optimization problem is supposed to be a scalar output?

Reply:
Yeah it does, I was also not sure if it would, but disabling `checks` allows it to work. It's a hack but I am not sure what other way is available?

Reply:
Let's open an issue about removing this. Constraints should probably have a way of being a NonlinearSystem for the constraint-satisfaction problem. In fact, `structural_simplify` should have the interpretation of tearing and simplification on that constraint-satisfaction system, i.e. if `x ~ y` is a constraint then it should just be eliminated from the system. What you could do is transform the OptimizationSystem to the NonlinearSystem of the constraints, apply tearing, and then grab the observed equations and reduced system, and make that be the reduced system of the OptimizationSystem. I'll chat with @YingboMa on how hard this would be, but I assume given what we have it could take like a day to implement, and this would be amazing.